src/macaron/database/table_definitions.py

```diff
@@ -1,128 +1,538 @@
+        p_name = "".join(c if c in allowed_chars else "_" for c in self.name)
+        return os.path.join(p_type, p_namespace, p_name)
+
+
+class Repository(ORMBase):
     """ORM Class for a repository."""
 
     __tablename__ = "_repository"
+
+    #: The primary key.
     id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)  # noqa: A003
+
+    # Because component is the parent table, we should define the foreign key here in the child table.
+    #: The foreign key to the software component.
+    component_id: Mapped[int] = mapped_column(Integer, ForeignKey("_component.id"))
+
+    #: The one-to-one relationship with the software component for this repository.
+    component: Mapped["Component"] = relationship(back_populates="repository", uselist=False, lazy="immediate")
+
+    # :Repository name in ``<git-service>/<owner>/<repo_name>`` format.
+    complete_name: Mapped[str] = mapped_column(String, nullable=False)
+
+    # :Repository name in ``<owner>/<repo_name>`` format.
     full_name: Mapped[str] = mapped_column(String, nullable=False)
-    remote_path: Mapped[str] = mapped_column(String, nullable=True)
+
+    #: The PURL type.
+    type: Mapped[str] = mapped_column(String, nullable=False)  # noqa: A003
+
+    # TODO: for locally cloned repos, do we have both type and owner, or can they be null?
+    #: The PURL namespace, which is the owner in pkg:github.com/owner/name@commit-sha.
+    owner: Mapped[str] = mapped_column(String, nullable=True)
+
+    #: The PURL name.
+    name: Mapped[str] = mapped_column(String, nullable=False)
+
+    #: The remote URL path to the repo.
+    remote_path: Mapped[str] = mapped_column(String, nullable=False)
+
+    #: The branch name.
     branch_name: Mapped[str] = mapped_column(String, nullable=False)
+
+    #: The release tag.
     release_tag: Mapped[str] = mapped_column(String, nullable=True)
+
+    #: The commit sha.
     commit_sha: Mapped[str] = mapped_column(String, nullable=False)
+
+    #: The commit date.
     commit_date: Mapped[str] = mapped_column(String, nullable=False)
 
+    #: The path to the repo on the file system.
+    fs_path: Mapped[str] = mapped_column(String, nullable=False)
+
+    def __init__(self, *args, files: list[str] | None = None, **kwargs):  # type: ignore[no-untyped-def]
+        """Instantiate the repository and set files.
+
+        Parameters
+        ----------
+        files: list[str] | None
+            The files extracted for this repository.
+        """
+        super().__init__(*args, **kwargs)
+
+        # We populate the PURL type, namespace, and name columns using the complete_name.
+        # Because locally cloned repositories may miss the namespace, we need to check the length.
+        # Note that the length of parts is already checked and should be either 2 or 3.
+        parts = [str(p) for p in Path(self.complete_name).parts]
```

Comment:
Why do we need to cast `p` to string even though `Path(...).parts` return a tuple of str elements :thinking: ?

Reply:
No it is not necessary. See commit [ba4f943](https://github.com/oracle/macaron/pull/305/commits/ba4f9439fc7f19b6d49dd78b4a8754ac5ef79394)